### PR TITLE
[13.0][ADD] Add new module sale_order_deliverable_rate

### PR DESCRIPTION
**Pull request by mathieuOdobescu**
This module adds a deliverable rate to a sale order. The deliverable rate is use to know (in percentage) how much of the order the stock can satisfy it.

### CODE REVIEW

File: sale_order_deliverable_rate/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            # we make avg of their deliverable_rate
+            if nb_order_line > 0:
+                rec.deliverable_rate = (
+                    sum(rec.order_line.mapped("deliverable_rate")) / nb_order_line
+                )
+            else:
+                rec.deliverable_rate = 100
+
+    def _search_deliverable_rate(self, operator, value):
+        lines = self.env["sale.order.line"].search([("qty_to_ship", ">=", 0)])
```
Comment (acsonefho) [inline sale_order_deliverable_rate/models/sale_order.py]: You can not search on no-stored field (cfr `sale.order.line` model).

And if you can use a more restrictive `domain`, it should be better for perf.

File: sale_order_deliverable_rate/models/sale_order_line.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    def _compute_qty_to_ship(self):
+        for rec in self:
+            rec.qty_to_ship = rec.product_uom_qty - rec.qty_delivered
+
+    @api.depends("product_uom_qty", "qty_delivered", "qty_to_ship")
+    def _compute_deliverable_rate(self):
+        for rec in self:
+            if rec.qty_to_ship > 0:
+                rate = rec.product_id.qty_available / rec.qty_to_ship
+                rec.deliverable_rate = 100 if rate >= 1 else rate * 100
```
Comment (acsonefho) [inline sale_order_deliverable_rate/models/sale_order_line.py]: More readable like this:
`rec.deliverable_rate = min(100, rate * 100)`

File: sale_order_deliverable_rate/tests/test_sale_order.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    def test_totally_deliverable(self):
+        """
+        Ensure that deliverable_rate of an order is to 100%
+        when it contains one order_line with a product who has enough stock
+        :return:
+        """
+        self._update_stock(self.array_products[0]["product"], 10)
+        self.array_products[0]["product_uom_qty"] = 10
+        so = self.create_basic_so(self.array_products[:1])
```
Comment (acsonefho) [inline sale_order_deliverable_rate/tests/test_sale_order.py]: You're creating a sale order into every test (except one). So best practice is to create it once during `setUpClass()`.
But you can let it into a dedicated function.

File: sale_order_deliverable_rate/tests/test_sale_order_line.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                "product": cls.Product.create(
+                    {"name": "Test Product 1", "type": "product"}
+                ),
+                "product_uom_qty": 0,
+                "qty_delivered": 0,
+            }
+        ]
+
+    @classmethod
+    def _update_stock(cls, product, qty):
```
Comment (acsonefho) [inline sale_order_deliverable_rate/tests/test_sale_order_line.py]: As there are some shared function/code into your 2 tests files, you have to create a `CommonClass`.

File: sale_order_deliverable_rate/views/sale_order.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2022 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record model="ir.ui.view" id="sale_order_form_view">
+        <field name="name">sale.order.form (in sale_order_deliverable_rate)</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form" />
```
Comment (acsonefho) [inline sale_order_deliverable_rate/views/sale_order.xml:8]: Missing priority.
`<field name="priority" eval="90" />`

File: sale_order_deliverable_rate/views/sale_order.xml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            >
+                <field name="deliverable_rate" optional="hide" />
+                <field name="qty_to_ship" invisible="1" />
+            </xpath>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="sale_order_tree_view">
+        <field name="name">sale.order.tree (in sale_order_deliverable_rate)</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_tree" />
```
Comment (acsonefho) [inline sale_order_deliverable_rate/views/sale_order.xml:29]: Missing priority.
`<field name="priority" eval="90" />`

File: sale_order_deliverable_rate/views/sale_order.xml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                    widget="progressbar"
+                    options="{'current_value': 'deliverable_rate', 'max_value': '100', 'editable': false}"
+                />
+            </xpath>
+        </field>
+    </record>
+    <record model="ir.ui.view" id="sale_order_search_view">
+        <field name="name">sale.order.search (in sale_order_deliverable_rate)</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_sales_order_filter" />
```
Comment (acsonefho) [inline sale_order_deliverable_rate/views/sale_order.xml:46]: Missing priority.
`<field name="priority" eval="90" />`

File: sale_order_deliverable_rate/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "sale order deliverable rate",
+    "version": "13.0.1.0.0",
+    "category": "Warehouse",
+    "summary": "Rate of deliverable product on a sale order",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV, Odoo Community Association (OCA)",
+    "website": "http://github.com/OCA/stock-logistics-workflow",
+    "depends": ["sale_stock", "product"],
```
Comment (acsonefho) [inline sale_order_deliverable_rate/__manifest__.py]: Useless `product` depends. Already in `sale_stock`. Not bad but useless.

File: sale_order_deliverable_rate/models/sale_order_line.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+    qty_to_ship = fields.Integer(compute="_compute_qty_to_ship", store=False)
+    deliverable_rate = fields.Integer(compute="_compute_deliverable_rate", store=False)
+
+    @api.depends("product_uom_qty", "qty_delivered")
+    def _compute_qty_to_ship(self):
+        for rec in self:
+            rec.qty_to_ship = rec.product_uom_qty - rec.qty_delivered
+
+    @api.depends("product_uom_qty", "qty_delivered", "qty_to_ship")
```
Comment (acsonefho) [inline sale_order_deliverable_rate/models/sale_order_line.py]: You can add `product_id` into your depends as you're using it.
But I don't know why you have `product_uom_qty` as you don't use it at all.

**Review by acsonefho — CHANGES_REQUESTED** [review]

**Review by acsonefho — APPROVED** [review]
Code LGTM

**Review by rousseldenis — APPROVED** [review]

File: sale_order_deliverable_rate/tests/test_sale_order.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        for product in self.array_products:
+            self._update_stock(product["product"], 0)
+
+        self.assertEqual(self.so_three_line.deliverable_rate, 0)
+
+    def test_search_deliverable_rate(self):
+        """
+        Ensure that the filter on deliverable_rate is finding order
+        """
+        orders = self.env["sale.order"].search([("deliverable_rate", "<", 100)])
```
Comment (acsonefho) [inline sale_order_deliverable_rate/tests/test_sale_order.py]: Your test is really not relevant... You must be more precise

**Review by acsonefho — CHANGES_REQUESTED** [review]

File: sale_order_deliverable_rate/tests/test_sale_order.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        first(self.so_one_line.order_line).write(
+            {"product_uom_qty": 10, "qty_delivered": 0}
+        )
+
+        self.assertEqual(self.so_one_line.deliverable_rate, 90)
+
+        orders = self.env["sale.order"].search([("deliverable_rate", ">=", 80)])
+
+        self.assertTrue(orders)
+        self.assertTrue(self.so_one_line in orders)
```
Comment (acsonefho) [inline sale_order_deliverable_rate/tests/test_sale_order.py]: `self.assertIn(self.so_one_line, orders)`

File: sale_order_deliverable_rate/tests/test_sale_order.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            self.assertGreaterEqual(order.deliverable_rate, 80)
+
+        # change the deliverable rate value by changing stock available
+        self._update_stock(self.array_products[0]["product"], 7)
+
+        self.assertEqual(self.so_one_line.deliverable_rate, 70)
+
+        orders = self.env["sale.order"].search([("deliverable_rate", ">=", 80)])
+        # check if the order is gone
+        self.assertFalse(self.so_one_line in orders)
```
Comment (acsonefho) [inline sale_order_deliverable_rate/tests/test_sale_order.py]: Same remark here

**Review by acsonefho — CHANGES_REQUESTED** [review]
Minor changes

**Review by acsonefho — APPROVED** [review]
Code LGTM

File: sale_order_deliverable_rate/models/sale_order.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        search="_search_deliverable_rate",
+        digits="Sale order deliverable rate",
+        store=False,
+    )
+
+    no_order_line_need_delivery = fields.Float(
+        compute="_compute_order_line_need_delivery", store=False
+    )
+
+    def order_line_need_delivery(self):
```
Comment (Cedric-Pigeon) [inline sale_order_deliverable_rate/models/sale_order.py]: a better name would be "_get_order_lines_to_ship"

File: sale_order_deliverable_rate/models/sale_order.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+    deliverable_rate = fields.Float(
+        compute="_compute_deliverable_rate",
+        search="_search_deliverable_rate",
+        digits="Sale order deliverable rate",
+        store=False,
+    )
+
+    no_order_line_need_delivery = fields.Float(
+        compute="_compute_order_line_need_delivery", store=False
```
Comment (Cedric-Pigeon) [inline sale_order_deliverable_rate/models/sale_order.py]: compute method must contain field name. "_compute_no_order_line_need_delivery"
"nb_" would be better to understand than "no_"

File: sale_order_deliverable_rate/models/sale_order.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    )
+
+    def order_line_need_delivery(self):
+        self.ensure_one()
+        return self.order_line.filtered(
+            lambda line: line.product_id.type == "product" and line.qty_to_ship > 0
```
Comment (Cedric-Pigeon) [inline sale_order_deliverable_rate/models/sale_order.py]: add "not line.display_type" to exclude sections and note lines

File: sale_order_deliverable_rate/models/sale_order.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+                    / rec.no_order_line_need_delivery
+                )
+            else:
+                # if everything is delivered deliverable rate => 0
+                rec.deliverable_rate = 0
+
+    def _search_deliverable_rate(self, operator, value):
+        lines = self.env["sale.order.line"].search([("qty_to_ship", ">=", 0)])
+
+        orders = lines.mapped("order_id")
```
Comment (Cedric-Pigeon) [inline sale_order_deliverable_rate/models/sale_order.py]: use a read-group for better performances:

order_data = self.env["sale.order.line"].read_group([("qty_to_ship", ">=", 0)], ["qty_to_ship"], ["order_id"])
order_ids = [data['order_id'][0] for data in order_data]
orders = self.browse(order_ids)

**Pull request merged**